pkg/gardenlet/controller/federatedseed/extensions/controllerinstallation_control.go

```diff
@@ -0,0 +1,213 @@
+			}
+
+			if !controllerRegistrationNamesForKind.Has(obj.Spec.RegistrationRef.Name) {
+				continue
+			}
+
+			c.controllerInstallationQueue.Add(obj.Name)
+		}
+
+		return reconcile.Result{}, nil
+	}
+}
+
+// createControllerInstallationRequiredReconcileFunc reconciles ControllerInstallations. It checks whether it is still
+// required by using the <kindToRequiredTypes> map that was computed by a separate reconciler.
+func (c *controllerInstallationControl) createControllerInstallationRequiredReconcileFunc(ctx context.Context) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		controllerInstallation := &gardencorev1beta1.ControllerInstallation{}
+		if err := c.k8sGardenClient.Client().Get(ctx, req.NamespacedName, controllerInstallation); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		controllerRegistration := &gardencorev1beta1.ControllerRegistration{}
+		if err := c.k8sGardenClient.Client().Get(ctx, kutil.Key(controllerInstallation.Spec.RegistrationRef.Name), controllerRegistration); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		var (
+			allKindsCalculated = true
+			required           *bool
+		)
+
+		c.lock.RLock()
+		for _, resource := range controllerRegistration.Spec.Resources {
+			requiredTypes, ok := c.kindToRequiredTypes[resource.Kind]
+			if !ok {
+				allKindsCalculated = false
+				continue
+			}
+
+			// at least one required type found
+			if requiredTypes.Has(resource.Type) {
+				required = pointer.BoolPtr(true)
+				break
+			}
+		}
+		c.lock.RUnlock()
+
+		if required == nil {
+			if !allKindsCalculated {
+				// if required wasn't set yet then but not all kinds were calculated then the it's not possible to
+				// decide yet whether it's required or not
+				return reconcile.Result{}, nil
+			}
+
+			// if required wasn't set yet then but all kinds were calculated then the installation is no longer required
+			required = pointer.BoolPtr(false)
+		}
+
+		if err := updateControllerInstallationRequiredCondition(ctx, c.k8sGardenClient.Client(), controllerInstallation, *required); err != nil {
+			c.log.Error(err)
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+}
+
+func updateControllerInstallationRequiredCondition(ctx context.Context, c client.Client, controllerInstallation *gardencorev1beta1.ControllerInstallation, required bool) error {
+	var (
+		conditionRequired = gardencorev1beta1helper.GetOrInitCondition(controllerInstallation.Status.Conditions, gardencorev1beta1.ControllerInstallationRequired)
+
+		status  = gardencorev1beta1.ConditionTrue
+		reason  = "ExtensionObjectsExist"
+		message = "extension objects exist in the seed having the kind/type combinations the controller is responsible for"
+	)
+
+	if !required {
+		status = gardencorev1beta1.ConditionFalse
+		reason = "NoExtensionObjects"
+		message = "no " + message
+	}
+
+	return kutil.TryUpdateStatus(ctx, retry.DefaultBackoff, c, controllerInstallation, func() error {
```

Comment:
I don't want to mix responsibilities here. There is the `ControllerRegistration` controller in the Gardener Controller Manager that is responsible for creating and deleting `ControllerInstallation`s at the right point in time. Especially with #2278 it will evaluate way more settings (e.g., deployment policy) that are relevant for the decision - not only if there are still extension resources pending.